ws/compression.nim

```diff
@@ -0,0 +1,212 @@
+## nim-ws
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+
+import
+  std/[strutils],
+  pkg/[stew/results, chronos],
+  ./types, ./frame, ./ext_utils, ./miniz/miniz_api
+
+type
+  DeflateOpts = object
+    isServer: bool
+    serverNoContextTakeOver: bool
+    clientNoContextTakeOver: bool
+    serverMaxWindowBits: int
+    clientMaxWindowBits: int
+
+  DeflateExt = ref object of Ext
+    paramStr: string
+    opts: DeflateOpts
+    # 'messageCompressed' is a two way flag:
+    # 1. the original message is compressible or not
+    # 2. the frame we received need decompression or not
+    messageCompressed: bool
+
+const
+  extID = "permessage-deflate"
+
+proc concatParam(resp: var string, param: string) =
+  resp.add ';'
+  resp.add param
+
+proc validateWindowBits(arg: ExtParam, res: var int): Result[string, string] =
+  if arg.value.len == 0:
+    return ok("")
+
+  if arg.value.len > 2:
+    return err("window bits expect 2 bytes, got " & $arg.value.len)
+
+  for n in arg.value:
+    if n notin Digits:
+      return err("window bits value contains illegal char: " & $n)
+
+  var winbit = 0
+  for i in 0..<arg.value.len:
+    winbit = winbit * 10 + arg.value[i].int - '0'.int
+
+  if winbit < 8 or winbit > 15:
+    return err("window bits should between 8-15, got " & $winbit)
+
+  res = winbit
+  return ok("=" & arg.value)
+
+proc validateParams(args: seq[ExtParam],
```

Comment:
Would something like `createParams` reflect it's intent better?

Reply:
fixed